vignettes/datatable-joins.Rmd

```diff
@@ -0,0 +1,687 @@
+- `vignette("datatable-keys-fast-subset", package="data.table")`
+
+***
+
+## 1. Defining example data
+
+To illustrate how to use the method available with real life examples, let's simulate a **normalized database** from a little supermarket by performing the following steps:
+
+1. Defining a `data.table` where each product is represented by a row with some qualities, but leaving one product without `id` to show how the framework deals with ***missing values***.
+
+```{r}
+Products = data.table(
+  id = c(1:4, 
+         NA_integer_),
+  name = c("banana",
+           "carrots",
+           "popcorn",
+           "soda",
+           "toothpaste"),
+  price = c(0.63,
+            0.89,
+            2.99,
+            1.49,
+            2.99),
+  unit = c("unit",
+           "lb",
+           "unit",
+           "ounce",
+           "unit"),
+  type = c(rep("natural", 2L),
+           rep("processed", 3L))
+)
+
+Products
+```
+
+2. Defining a `data.table` showing the proportion of taxes to be applied for processed products based on their units.
+
+```{r}
+NewTax = data.table(
+  unit = c("unit","ounce"),
+  type = "processed",
+  tax_prop = c(0.65, 0.20)
+)
+
+NewTax
+```
+
+
+3. Defining a `data.table` simulating the products received every Monday with a `product_id` that is not present in the `Products` table.
+
+```{r}
+set.seed(2156)
+
+ProductReceived = data.table(
+  id = 1:10,
+  date = seq(from = as.IDate("2024-01-08"), length.out = 10L, by = "week"), 
+  product_id = sample(c(NA_integer_, 1:3, 6L), size = 10L, replace = TRUE),
+  count = sample(c(50L, 100L, 150L), size = 10L, replace = TRUE)
+)
+
+ProductReceived
+```
+
+4. Defining a `data.table` to show some sales that can take place on weekdays with another `product_id` that is not present in the `Products` table.
+
+```{r}
+sample_date = function(from, to, size, ...){
+  all_days = seq(from = from, to = to, by = "day")
+  weekdays = all_days[wday(all_days) %in% 2:6]
+  days_sample = sample(weekdays, size, ...)
+  days_sample_desc = sort(days_sample)
+  return(days_sample_desc)
+}
+
+set.seed(5415)
+
+ProductSales = data.table(
+  id = 1:10,
+  date = ProductReceived[, sample_date(min(date), max(date), 10L)], 
+  product_id = sample(c(1:3, 7L), size = 10L, replace = TRUE),
+  count = sample(c(50L, 100L, 150L), size = 10L, replace = TRUE)
+)
+
+
+ProductSales
+```
+
+## 2. `data.table` joining syntax 
+
+Before taking advantage of the `data.table` syntax to perform join operations we need to know which arguments can help us to perform successful joins.
+
+The next diagram shows a description for each basic argument. In the following sections we will show how to use each of them and add more complexity little by little.
+
+```
+x[i, on, nomatch]
+| |  |   |
+| |  |   ----> If NULL only returns rows linked in x and i tables
+| |  ----> a character vector o list defining match logict
+| ----> principal data.table, list or data.frame
```

Comment:
primary instead of principal?

also where do these names, principal/secondary come from? I don't see them in ?data.table so I would suggest to keep the standard/docmented names, `i` and `x`

Reply:
Toby, thanks for pointing this out, I will change it from principal to primary.

I am trying to keep using the documented reference.

I just thought it could be useful for a new user to understand that the `i` table is more important in defining the number of rows to return, regardless of whether they are trying to make a left or right join.